py/binary.c

```diff
@@ -478,16 +483,12 @@ void mp_binary_set_val_array(char typecode, void *p, size_t index, mp_obj_t val_
         case 'O':
             ((mp_obj_t *)p)[index] = val_in;
             break;
-        default:
-            #if MICROPY_LONGINT_IMPL != MICROPY_LONGINT_IMPL_NONE
-            if (mp_obj_is_exact_type(val_in, &mp_type_int)) {
-                size_t size = mp_binary_get_size('@', typecode, NULL);
-                mp_obj_int_to_bytes_impl(val_in, MP_ENDIANNESS_BIG,
-                    size, (uint8_t *)p + index * size);
-                return;
-            }
-            #endif
-            mp_binary_set_val_array_from_int(typecode, p, index, mp_obj_get_int(val_in));
+        default: {
+            size_t size = mp_binary_get_size('@', typecode, NULL);
+            p = (uint8_t *)p + index * size;
+            mp_obj_int_to_bytes(val_in, size, p, MP_ENDIANNESS_BIG, is_signed(typecode), OVERFLOW_CHECKS);
```

Comment:
This endianness choice seems wrong, though, at least `mp_binary_set_val_array_from_int` would have used native endianness...? Need to make sure this is being tested.

Reply:
This is fine, I misread the meaning of `MP_ENDIANNESS_BIG` here as being an argument to the function, but the argument is `bool big_endian` and the value of `MP_ENDIANNESS_BIG` macro is 0 on a little endian port.